yamlfile.c

```diff
@@ -305,29 +159,116 @@ StatusCode fiftyoneDegreesYamlFileIterateWithLimit(
 	int limit,
 	void* state,
 	void(*callback)(KeyValuePair*, uint16_t, void*)) {
+	char* current;
 	FILE *handle;
 	int count = 0;
-
 	StatusCode status = FileOpen(fileName, &handle);
-	FileState fileState = { handle, buffer, 0, 0, length, 0, 0 };
-	if (status == SUCCESS) {
-		// Read the first block
-		status = readNextBlock(&fileState);
-		if (status == SUCCESS ||
-			status == FILE_END_OF_FILE) {
-			while ((count < limit || limit < 0)
-				&& (status == SUCCESS || status == FILE_END_OF_DOCUMENT)) {
-				status = readNextDocument(
-					&fileState,
-					keyValuePairs,
-					collectionSize,
-					state,
-					callback);
-				count++;
+	if (status != SUCCESS) {
+		return status;
+	}
+
+	FileState fileState = { 
+		handle, 
+		buffer, 
+		length,
+		// Set the end and current to 0 to force the next block to be read.
+		0,
+		0,
+		0, 
+		0,
+		0,
+		0,
+		0,
+		false };
+
+	PairState pairState = {
+		keyValuePairs,
+		collectionSize,
+		0,
+		keyValuePairs[0].key,
+		keyValuePairs[0].key + keyValuePairs[0].keyLength - 1 };
+
+	// If there is no limit then set the limit to the largest value to 
+	// avoid checking for negative values in the loop.
+	if (limit < 0) {
+		limit = INT_MAX;
+	}
+
+	while (*(current = readNext(&fileState)) != '\0') {
+
+		// If the first new line then move to the next pair.
+		if (*current == '\n' || *current == '\r') {
+			if (fileState.newLineCount == 0) {
+
+				// If there was a quote earlier in the string and the last one
+				// is also a quote then remove the last quote.
+				if (fileState.quote > 0 && 
+					*(pairState.current - 1) == '\'') {
+					pairState.current--;
+				}
+				nextPair(&pairState);
 			}
+			newLine(&fileState);
+			fileState.newLineCount++;
+		}
+
+		// If a dash and control length is reached then return the pairs
+		// and reset ready for the next set.
+		else if (*current == '-' && fileState.position == fileState.dashCount) {
+			advance(&fileState);
+			fileState.dashCount++;
+			if (fileState.dashCount == CONTROL_LENGTH) {
+				if (pairState.index > 0)
+				{
+					callback(keyValuePairs, pairState.index, state);
+					count++;
+					if (count >= limit) {
+						break;
+					}
+				}
+				newDocument(&pairState);
+			}
+		}
+
+		// If a dot and control length then return and exit.
+		else if (*current == '.' && fileState.position == fileState.dotCount) {
+			advance(&fileState);
+			fileState.dotCount++;
+			if (fileState.dotCount == CONTROL_LENGTH) {
+				callback(keyValuePairs, pairState.index, state);
```

Comment:
This branch (`...`) does not check `pairState.pairIndex` unlike the one above (`---`).

So in case of
```yaml
---
...
```
the empty "document" will be reported
while in
```yaml
---
---
```
it is not.

Please see https://github.com/51Degrees/common-cxx/pull/41 for added test cases and a patch.